src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1046 @@
+int lsrkStep_AccessARKODEStepMem(void* arkode_mem, const char* fname,
+                                 ARKodeMem* ark_mem, ARKodeLSRKStepMem* step_mem)
+{
+  /* access ARKodeMem structure */
+  if (arkode_mem == NULL)
+  {
+    arkProcessError(NULL, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_ARK_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  *ark_mem = (ARKodeMem)arkode_mem;
+
+  /* access ARKodeLSRKStepMem structure */
+  if ((*ark_mem)->step_mem == NULL)
+  {
+    arkProcessError(*ark_mem, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_LSRKSTEP_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  *step_mem = (ARKodeLSRKStepMem)(*ark_mem)->step_mem;
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_AccessStepMem:
+
+  Shortcut routine to unpack the step_mem structure from
+  ark_mem.  If missing it returns ARK_MEM_NULL.
+  ---------------------------------------------------------------*/
+int lsrkStep_AccessStepMem(ARKodeMem ark_mem, const char* fname,
+                           ARKodeLSRKStepMem* step_mem)
+{
+  /* access ARKodeLSRKStepMem structure */
+  if (ark_mem->step_mem == NULL)
+  {
+    arkProcessError(ark_mem, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_LSRKSTEP_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  *step_mem = (ARKodeLSRKStepMem)ark_mem->step_mem;
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_CheckNVector:
+
+  This routine checks if all required vector operations are
+  present.  If any of them is missing it returns SUNFALSE.
+  ---------------------------------------------------------------*/
+sunbooleantype lsrkStep_CheckNVector(N_Vector tmpl)
+{
+  if ((tmpl->ops->nvclone == NULL) || (tmpl->ops->nvdestroy == NULL) ||
+      (tmpl->ops->nvlinearsum == NULL) || (tmpl->ops->nvconst == NULL) ||
+      (tmpl->ops->nvscale == NULL) || (tmpl->ops->nvwrmsnorm == NULL))
+  {
+    return (SUNFALSE);
+  }
+  return (SUNTRUE);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_SprRadUpdateLogic:
+
+  This routine checks if the step is accepted or not and reassigns
+  the SprRad update flags accordingly.
+  ---------------------------------------------------------------*/
+
+int lsrkStep_ComputeNewSprRad(ARKodeMem ark_mem, ARKodeLSRKStepMem step_mem)
+{
+  int retval = SUN_SUCCESS;
+
+  if ((step_mem->isextspr))
+  {
+    retval = step_mem->extspr(ark_mem->tn, &step_mem->sprad, ark_mem->user_data);
+    step_mem->sprad *= step_mem->sprsfty;
+    step_mem->sprad = SUNRabs(step_mem->sprad);
+  }
+  else
+  {
+    printf("\nInternal SprRad is not supported yet!");
+    printf(
+      "\nCall LSRKodeSetSprRadFn to provide an external SprRad function\n");
+
+    return (ARK_ILL_INPUT);
+  }
+  step_mem->jacatt = SUNTRUE;
+
+  step_mem->sprmax = (round(step_mem->sprad) > step_mem->sprmax)
```

Comment:
Do you really want `round` or would `ceil` be more appropriate?  Also, we may need precision-aware SUNDIALS macros instead of relying on `round` (that may just return a `double`).